File: clients/kotlin_graal/src/main/kotlin/komus24/model/DebugState.kt
package komus24.model

import komus24.util.StreamUtil

/**
 * TODO - Document
 */
class DebugState {
    /**
     * TODO - Document
     */
    var pressedKeys: Array<String>

    constructor(pressedKeys: Array<String>) {
        this.pressedKeys = pressedKeys
    }

    /**
     * Write DebugState to output stream
     */
    @Throws(java.io.IOException::class)
    fun writeTo(stream: java.io.OutputStream) {
        StreamUtil.writeInt(stream, pressedKeys.size)
        for (pressedKeysElement in pressedKeys) {
            StreamUtil.writeString(stream, pressedKeysElement)
        }
    }

    /**
     * Get string representation of DebugState
     */
    override fun toString(): String {
        var stringBuilder = StringBuilder("DebugState { ")
        stringBuilder.append("pressedKeys: ")
        stringBuilder.append("[ ")
        var pressedKeysIndex = 0
        for (pressedKeysElement in pressedKeys) {
            if (pressedKeysIndex != 0) {
                stringBuilder.append(", ")
            }
            stringBuilder.append('"' + pressedKeysElement + '"')
            pressedKeysIndex++
        }
        stringBuilder.append(" ]")
        stringBuilder.append(" }")
        return stringBuilder.toString()
    }

    companion object {
        /**
         * Read DebugState from input stream
         */
        @Throws(java.io.IOException::class)
        fun readFrom(stream: java.io.InputStream): DebugState {
            var pressedKeys: Array<String>
            pressedKeys = Array(StreamUtil.readInt(stream), {
                var pressedKeysElement: String
                pressedKeysElement = StreamUtil.readString(stream)
                pressedKeysElement
            })
            return DebugState(pressedKeys)
        }
    }
}

File: clients/cpp_clang/model/DebugState.cpp
#include "DebugState.hpp"

namespace model {

DebugState::DebugState(std::vector<std::string> pressedKeys) : pressedKeys(pressedKeys) { }

// Read DebugState from input stream
DebugState DebugState::readFrom(InputStream& stream) {
    std::vector<std::string> pressedKeys = std::vector<std::string>();
    size_t pressedKeysSize = stream.readInt();
    pressedKeys.reserve(pressedKeysSize);
    for (size_t pressedKeysIndex = 0; pressedKeysIndex < pressedKeysSize; pressedKeysIndex++) {
        std::string pressedKeysElement = stream.readString();
        pressedKeys.emplace_back(pressedKeysElement);
    }
    return DebugState(pressedKeys);
}

// Write DebugState to output stream
void DebugState::writeTo(OutputStream& stream) const {
    stream.write((int)(pressedKeys.size()));
    for (const std::string& pressedKeysElement : pressedKeys) {
        stream.write(pressedKeysElement);
    }
}

// Get string representation of DebugState
std::string DebugState::toString() const {
    std::stringstream ss;
    ss << "DebugState { ";
    ss << "pressedKeys: ";
    ss << "[ ";
    for (size_t pressedKeysIndex = 0; pressedKeysIndex < pressedKeys.size(); pressedKeysIndex++) {
        const std::string& pressedKeysElement = pressedKeys[pressedKeysIndex];
        if (pressedKeysIndex != 0) {
            ss << ", ";
        }
        ss << '"' << pressedKeysElement << '"';
    }
    ss << " ]";
    ss << " }";
    return ss.str();
}

}